Source/HTTPHeaders.swift

```diff
@@ -24,51 +24,215 @@
+    }
+    
+    public mutating func update(name: String, value: String) {
+        guard let index = headers.index(of: name) else {
+            headers.append(HTTPHeader(name: name, value: value))
+            return
+        }
+        
+        headers.replaceSubrange(index...index, with: [HTTPHeader(name: name, value: value)])
+    }
+    
+    public mutating func update(_ header: HTTPHeader) {
+        update(name: header.name, value: header.value)
+    }
+    
+    public mutating func remove(name: String) {
+        guard let index = headers.index(of: name) else { return }
+        
+        headers.remove(at: index)
+    }
+    
+    mutating func sort() {
+        return headers.sort()
+    }
+    
+    func sorted() -> HTTPHeaders {
+        return HTTPHeaders(headers.sorted())
+    }
+    
+    func value(for name: String) -> String? {
+        guard let index = headers.index(of: name) else { return nil }
+        
+        return headers[index].value
+    }
+    
+    public subscript(_ name: String) -> String? {
+        get { return value(for: name) }
+        set {
+            if let value = newValue {
+                update(name: name, value: value)
+            } else {
+                remove(name: name)
+            }
+        }
+    }
+}
+
+extension HTTPHeaders: ExpressibleByDictionaryLiteral {
+    public init(dictionaryLiteral elements: (String, String)...) {
+        self.init()
+        
+        elements.forEach { update(name: $0.0, value: $0.1) }
+    }
+    
+    public var dictionary: [String: String] {
+        let namesAndValues = headers.map { ($0.name, $0.value) }
+        
+        return Dictionary(namesAndValues, uniquingKeysWith: { (_, last) in last })
+    }
+}
+
+extension HTTPHeaders: ExpressibleByArrayLiteral {
+    public init(arrayLiteral elements: HTTPHeader...) {
+        self.init()
+        
+        elements.forEach { update(name: $0.name, value: $0.value) }
+    }
+}
+
+extension HTTPHeaders: Sequence {
+    public func makeIterator() -> IndexingIterator<Array<HTTPHeader>> {
+        return headers.makeIterator()
+    }
+}
+
+extension HTTPHeaders: Collection {
+    public var startIndex: Int {
+        return headers.startIndex
+    }
+    
+    public var endIndex: Int {
+        return headers.endIndex
+    }
+    
+    public subscript(position: Int) -> HTTPHeader {
+        return headers[position]
+    }
+    
+    public func index(after i: Int) -> Int {
+        return headers.index(after: i)
+    }
+}
+
+extension HTTPHeaders: CustomStringConvertible {
+    public var description: String {
+        return headers.sorted().map { $0.description }.joined(separator: "\n")
+    }
+}
+
+// MARK: - HTTPHeader
+
+public struct HTTPHeader: Hashable {
+    public let name: String
+    public let value: String
+}
+
+extension HTTPHeader: CustomStringConvertible {
+    public var description: String {
+        return "\(name): \(value)"
+    }
+}
+
+extension HTTPHeader: Comparable {
```

Comment:
Not sure if it's entirely correct to conform `HTTPHeader` to `Comparable` if we're not also comparing the value. Since we're only sorting headers in one method, it might be better to use a custom sorting function (See line 62)

Reply:
Ah, because the synthesized `Equatable` does include `value`? I don't see an issue with including it here too, even though `name` should be enough.

Reply:
Yes, synthesized `Equatable` compares both values, but also, you're assuming that *any* collection of `HTTPHeader` will be free of duplicate names.
Consider this test:

```swift
let headersArr = [
    HTTPHeader(name: "foo", value: "a"),
    HTTPHeader(name: "foo", value: "b"),
    HTTPHeader(name: "foo", value: "c"),
]

var headers1 = HTTPHeaders()
var headers2 = HTTPHeaders()
var headers3 = HTTPHeaders()

for header in headersArr {
    headers1.update(header)
}

for header in headersArr.sorted() {
    headers2.update(header)
}

for header in headersArr.reversed().sorted() {
    headers3.update(header)
}
```

Currently, the result is:
```
headers1: ["foo": "c"]
headers2: ["foo": "c"]
headers3: ["foo": "a"]
```

The expected result would be:
```
headers1: ["foo": "c"]
headers2: ["foo": "c"]
headers3: ["foo": "c"]
```
... with the following change:
```swift
extension HTTPHeader : Comparable {
    public static func <(lhs: HTTPHeader, rhs: HTTPHeader) -> Bool {
        return lhs.name == rhs.name ? lhs.value < rhs.value : lhs.name < rhs.name
    }
}
```

Alternatively, since `HTTPHeaders` guarantees that names will be unique, we could remove `Comparable` conformance and change the `sort` function to:
```swift
mutating func sort() {
    return headers.sort{ $0.name < $1.name }
}
```

Reply:
You're right, and I've made the change for manual comparison.